### PR TITLE
{2023.06} foss/2022b

### PR DESCRIPTION
**Pull request by casparvl**

### CODE REVIEW

**Comment by casparvl** [comment]
bot: build repo:eessi-2023.06-software arch:x86_64/generic

**Comment by casparvl** [comment]
bot: build repo:eessi-2023.06-software arch:x86_64/generic

**Comment by casparvl** [comment]
bot: build repo:eessi-2023.06-software arch:x86_64/generic

**Comment by casparvl** [comment]
bot: build repo:eessi-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi-2023.06-software arch:aarch64/generic
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_n1

**Comment by casparvl** [comment]
The failure on `neoverse_v1`:

```
0:42:15  10 out of 14 easyconfigs done: BLIS/0.9.0-GCC-12.2.0 (OK), Python/3.10.ERROR: Build of /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/aarch64/neoverse_v1/software/EasyBuild/4.7.2/easybuild/easyconfigs/o/OpenBLAS/OpenBLAS-0.3.21-GCC-12.2.0.eb failed (err: 'build failed (first 300 chars): Too many LAPACK tests failed due to numerical errors: 344 (> 300)')
```

**Comment by casparvl** [comment]
I don't know either. I guess we can, but it does make one wonder: if numerical results are different, how different are they, and is that still acceptable? We should at the very least check that all failures are numerical failures, I guess.

I think the more fundamental question is: what should are tests guarantee?

1. If they should only guarantee that the _installation_ went ok, I think we can ignore the numerical errors: this is simply the behavior of _this_ OpenBLAS version on `neoverse_v1`. "Fixing" that is not up to us.
2. If they should guarantee that software produces the right result, we should _not_ deploy it on `neoverse_v1`, and only deploy a version there once these numerical inconsistencies have been resolved (which it might never be if the OpenBLAS devs don't consider it a prolbem).

It's probably good to at least report an issue upstream, see what they say. I'd assume the devs are more adept at judging whether these numerical inconsistencies should be considered problematic or not. One issue with taking the first approach is that issues may also pop up in other packages, such as https://github.com/EESSI/software-layer/pull/306

**Comment by casparvl** [comment]
More detailed error:

```
                        -->   LAPACK TESTING SUMMARY  <--
SUMMARY                 nb test run     numerical error         other error
================        ===========     =================       ================
REAL                    1315683         107     (0.008%)        0       (0.000%)
DOUBLE PRECISION        1314777         66      (0.005%)        0       (0.000%)
COMPLEX                 773609          97      (0.013%)        0       (0.000%)
COMPLEX16               776246          74      (0.010%)        0       (0.000%)

--> ALL PRECISIONS      4180315         344     (0.008%)        0       (0.000%)


== 2023-07-19 12:16:15,506 openblas.py:113 INFO 4180315 LAPACK tests run - 344 failed due to numerical errors - 0 failed due to other errors
```

**Comment by casparvl** [comment]
I managed to go into the singularity container by unpacking the tarball at `/mnt/shared/home/bot/eessi-bot-software-layer/jobs/2023.07/pr_309/5990/previous_tmp`

doing a `singularity shell ghcr.io_eessi_build_node_debian11.sif`

and find the detailed testing results at `bot/easybuild/build/OpenBLAS/0.3.21/GCC-12.2.0/OpenBLAS-0.3.21/lapack-netlib/TESTING/testing_results.txt`. An excerpt:

```
...
Matrix order=   15, type=10, seed= 798,1691,2423, 745, result  5 is 8.389E+06
 Matrix order=   15, type=11, seed= 931,1787, 557,2429, result  5 is 8.389E+06
 Matrix order=   15, type=17, seed= 529,3615,1764,1221, result  5 is 8.389E+06
 Matrix order=   15, type=18, seed=3991, 625,3539,2581, result  5 is 8.389E+06
 Matrix order=   15, type=19, seed=2400,1821, 218,1365, result  5 is 8.389E+06
 Matrix order=   15, type=20, seed=2155,2073,3686, 149, result  5 is 8.389E+06
 Matrix order=   15, type=21, seed=3823,2194,3510,3029, result  5 is 8.389E+06
 Matrix order=   15, type=22, seed=2951, 608,1256,3481, result  5 is 8.389E+06
 Matrix order=   15, type=23, seed=1243,2927, 263, 357, result  5 is 8.389E+06
 Matrix order=   15, type=24, seed=1487,3956,2976,3649, result  5 is 8.389E+06
 Matrix order=   15, type=25, seed=2605,2211,3982,2721, result  5 is 8.389E+06
 Matrix order=   15, type=26, seed=   8,1600,1726,2817, result  5 is 8.389E+06
 Matrix order=   20, type= 8, seed= 624, 693,1681,   9, result  5 is 8.389E+06
 Matrix order=   20, type=10, seed= 563,2705, 428, 633, result  5 is 8.389E+06
 Matrix order=   20, type=12, seed= 112,1059,2403,2121, result  5 is 8.389E+06
 Matrix order=   20, type=13, seed=3448,3524, 410,2697, result  5 is 8.389E+06
 Matrix order=   20, type=14, seed= 248,   6,3354,3273, result  5 is 8.389E+06
 Matrix order=   20, type=17, seed= 723,4024,2193,2265, result  5 is 8.389E+06
 Matrix order=   20, type=18, seed= 828, 842,1787,3733, result  5 is 8.389E+06
 Matrix order=   20, type=19, seed= 252,3327,3300,1389, result  5 is 8.389E+06
 Matrix order=   20, type=20, seed=1904,2346,3199,2437, result  5 is 8.389E+06
 Matrix order=   20, type=21, seed=4066,  26,1253, 221, result  5 is 8.389E+06
 Matrix order=   20, type=22, seed=3234,3664,3237,3473, result  5 is 8.389E+06
 Matrix order=   20, type=23, seed= 481, 957,3230,3689, result  5 is 8.389E+06
 Matrix order=   20, type=24, seed=2871,2618,1987,2949, result  5 is 8.389E+06
 Matrix order=   20, type=25, seed=2604,3233, 644,3529, result  5 is 8.389E+06
 Matrix order=   20, type=26, seed=1465,2266,3976,3833, result  5 is 8.389E+06
 SGV drivers:     64 out of   1092 tests failed to pass the threshold
...
 Matrix order=   20, type=10, seed= 563,2705, 428, 633, result  5 is 4.504D+15
 Matrix order=   20, type=12, seed= 112,1059,2403,2121, result  5 is 4.504D+15
 Matrix order=   20, type=13, seed=3448,3524, 410,2697, result  5 is 4.504D+15
 Matrix order=   20, type=17, seed= 723,4024,2193,2265, result  5 is 4.504D+15
 Matrix order=   20, type=18, seed= 828, 842,1787,3733, result  5 is 4.504D+15
 Matrix order=   20, type=19, seed= 252,3327,3300,1389, result  5 is 4.504D+15
 Matrix order=   20, type=20, seed=1904,2346,3199,2437, result  5 is 4.504D+15
 Matrix order=   20, type=21, seed=4066,  26,1253, 221, result  5 is 4.504D+15
 Matrix order=   20, type=22, seed=3234,3664,3237,3473, result  5 is 4.504D+15
 Matrix order=   20, type=23, seed= 481, 957,3230,3689, result  5 is 4.504D+15
 Matrix order=   20, type=24, seed=2871,2618,1987,2949, result  5 is 4.504D+15
 Matrix order=   20, type=25, seed=2604,3233, 644,3529, result  5 is 4.504D+15
 Matrix order=   20, type=26, seed=1465,2266,3976,3833, result  5 is 4.504D+15
 DGV drivers:     34 out of   1092 tests failed to pass the threshold
...
 Matrix order=    6, type=16, seed= 852,1170,4018,2529, result  5 is 4.504D+15
 Matrix order=    8, type=20, seed= 202,3818,2830,2373, result  5 is 4.504D+15
 Matrix order=    8, type=25, seed= 582,1680, 289,1745, result  5 is 4.504D+15
 Matrix order=   10, type= 8, seed= 573,2931,1057, 529, result  5 is 4.504D+15
 Matrix order=   10, type=10, seed= 308, 212,3333,  73, result  5 is 4.504D+15
 Matrix order=   10, type=13, seed=3036,2372,2403,2421, result  5 is 4.504D+15
 Matrix order=   15, type=12, seed=2288,1863,3505, 805, result  5 is 4.504D+15
 Matrix order=   15, type=13, seed=3415,2880,2965,2889, result  5 is 4.504D+15
 Matrix order=   15, type=16, seed= 681,2778,2914,2225, result  5 is 4.504D+15
 Matrix order=   15, type=17, seed= 304,3863, 350,1657, result  5 is 4.504D+15
 Matrix order=   15, type=21, seed=2362, 980, 534,3365, result  5 is 4.504D+15
 Matrix order=   15, type=25, seed=2888,1855, 900,2225, result  5 is 4.504D+15
 Matrix order=   20, type= 8, seed= 636, 119,  47,2269, result  5 is 4.504D+15
 Matrix order=   20, type= 9, seed=1007,2268,2913,  29, result  5 is 4.504D+15
 Matrix order=   20, type=10, seed= 104,1378,2159,1485, result  5 is 4.504D+15
 Matrix order=   20, type=11, seed= 927, 874,3720,2349, result  5 is 4.504D+15
 Matrix order=   20, type=14, seed=2659,2147, 552,1661, result  5 is 4.504D+15
 Matrix order=   20, type=21, seed=2012,3064,3728,2789, result  5 is 4.504D+15
 DGV drivers:     18 out of   1092 tests failed to pass the threshold
...
 N=  30 M=   0, P=   5, type  8, test  2, ratio=  555331.
 N=   3 M=   0, P=  20, type  8, test  2, ratio= 0.889025E+07
 N=  30 M=   0, P=  20, type  8, test  2, ratio= 0.219820E+07
 M=   3 P=   0, N=  30, type  7, test  1, ratio= 0.800042E+07
 N=  30 M=   3, P=   0, type  7, test  1, ratio= 0.772279E+07
 M=   3 P=   5, N=  30, type  7, test  1, ratio= 0.852669E+07
 N=  30 M=   3, P=   5, type  8, test  2, ratio= 0.119398E+07
 N=   3 M=   3, P=  20, type  8, test  2, ratio= 0.998240E+07
 N=  30 M=   3, P=  20, type  8, test  2, ratio=  976876.
 N=   3 M=  10, P=   5, type  8, test  2, ratio= 0.121475E+08
 N=  30 M=  10, P=   5, type  7, test  1, ratio= 0.127652E+08
 N=  30 M=  10, P=   5, type  8, test  2, ratio=  714258.
 M=  10 P=  20, N=  30, type  7, test  1, ratio= 0.510405E+07
 N=  30 M=  10, P=  20, type  8, test  2, ratio= 0.208493E+07
 GQR:     14 out of   1728 tests failed to pass the threshold
...
 Matrix order=   20, type=16, seed= 647,2328,1944, 557, result  5 is 8.389E+06
 CGV drivers:      1 out of   1092 tests failed to pass the threshold
...
 CHESV , UPLO='U', N =    3, type  2, test  1, ratio = 0.10874E+07
 CHESVX, FACT='N', UPLO='U', N =    3, type  2, test  1, ratio = 0.10874E+07
 CHESV , UPLO='U', N =    5, type  9, test  1, ratio = 0.32042E+07
 CHESVX, FACT='N', UPLO='U', N =    5, type  9, test  1, ratio = 0.32042E+07
 CHESV , UPLO='U', N =    5, type 10, test  1, ratio = 0.35394E+07
 CHESVX, FACT='N', UPLO='U', N =    5, type 10, test  1, ratio = 0.35394E+07
 CHESV , UPLO='U', N =   10, type  2, test  1, ratio = 0.16458E+07
 CHESVX, FACT='N', UPLO='U', N =   10, type  2, test  1, ratio = 0.16458E+07
 CHESV , UPLO='U', N =   10, type  7, test  1, ratio = 0.14266E+06
 CHESVX, FACT='N', UPLO='U', N =   10, type  7, test  1, ratio = 0.14266E+06
 CHESV , UPLO='U', N =   10, type  9, test  1, ratio = 0.31372E+07
 CHESVX, FACT='N', UPLO='U', N =   10, type  9, test  1, ratio = 0.31372E+07
 CHESV , UPLO='U', N =   10, type 10, test  1, ratio = 0.16814E+07
 CHESVX, FACT='N', UPLO='U', N =   10, type 10, test  1, ratio = 0.16814E+07
 CHESV , UPLO='U', N =   50, type  2, test  1, ratio = 0.86168E+07
 CHESVX, FACT='N', UPLO='U', N =   50, type  2, test  1, ratio = 0.86168E+07
 CHESV , UPLO='U', N =   50, type  7, test  1, ratio = 0.80869E+06
 CHESVX, FACT='N', UPLO='U', N =   50, type  7, test  1, ratio = 0.80869E+06
 CHESV , UPLO='U', N =   50, type  8, test  1, ratio = 0.52818E+06
 CHESVX, FACT='N', UPLO='U', N =   50, type  8, test  1, ratio = 0.52818E+06
 CHESV , UPLO='U', N =   50, type  9, test  1, ratio = 0.85655E+07
 CHESVX, FACT='N', UPLO='U', N =   50, type  9, test  1, ratio = 0.85655E+07
 CHESV , UPLO='U', N =   50, type 10, test  1, ratio = 0.32095E+07
 CHESVX, FACT='N', UPLO='U', N =   50, type 10, test  1, ratio = 0.32095E+07
 CHE drivers:     24 out of   1072 tests failed to pass the threshold
...
 Matrix order=   12, type=24, seed=2693,2404,3046,2957, result  5 is 4.504D+15
 Matrix order=   12, type=25, seed= 578, 723, 929,1637, result  5 is 4.504D+15
 Matrix order=   12, type=26, seed=2061,1512,1968, 125, result  5 is 4.504D+15
 Matrix order=   20, type=16, seed= 647,2328,1944, 557, result  5 is 4.504D+15
 Matrix order=   20, type=17, seed=1585,3902,3906,1293, result  5 is 4.504D+15
 Matrix order=   20, type=18, seed=1063,2113,3640,2685, result  5 is 4.504D+15
 Matrix order=   20, type=19, seed=2305,3879, 305, 381, result  5 is 4.504D+15
 Matrix order=   20, type=20, seed= 897,3616, 121,2173, result  5 is 4.504D+15
 Matrix order=   20, type=21, seed= 676,3851,3089,3965, result  5 is 4.504D+15
 Matrix order=   20, type=22, seed=1763,2855,1954,1469, result  5 is 4.504D+15
 Matrix order=   20, type=23, seed=1338,3437,3180,3285, result  5 is 4.504D+15
 Matrix order=   20, type=24, seed=3522,1685,3785, 813, result  5 is 4.504D+15
 Matrix order=   20, type=25, seed=2931,3978,3195,3781, result  5 is 4.504D+15
 Matrix order=   20, type=26, seed= 449,1038,1670,1437, result  5 is 4.504D+15
 ZGV drivers:     53 out of   1092 tests failed to pass the threshold
```

Now, I'm no expert in these tests, and have no idea what this means. It does not look like small numerical errors to my untrained eye, but I might be completely wrong...

**Comment by boegel** [comment]
@casparvl To avoid getting stuck on this even longer, I think we should:

1) Bump the limit for failing numerical tests for OpenBLAS to `350` or `400` in our hooks (see [here](https://github.com/EESSI/software-layer/blob/231a0054301945532a5a448c2e56caf03e2fa82f/eb_hooks.py#L162)), so the test step passes. The LAPACK test suite summary clearly show that just 0.008% of all (numerical) tests fail, I strongly feel that's not serious enough to block the installation, especially since this only happens for `aarch64/neoverse_v1`, and that we'll have a clear record of this in the EasyBuild log file that is included in the installation.

2) Open an issue in this repository to keep track of the fact that we're seeing a bit more numerical failures for OpenBLAS on `aarch64` than we do on `amd` or `intel` systems, with all the details (like the stuff you mentioned in your last comment, since this PR is not the right place to have this discussion imho), and also compare things across easyconfig generations/toolchains.

3) Report this upstream to [OpenBLAS](https://github.com/xianyi/OpenBLAS/issues), with sufficient detail, and keep track of the feedback in our own issue - we could probably use some help from @bartoldeman here if he can find the time for it, since he has some experience there.

We should take a similar approach to unblock other PRs (`numpy` in #306, FFTW in #297 and #310), and push on to discover other issues which no doubt will pop up.

We can't reasonable expect that we'll figure out each and every failing test for all software we'll install, especially because we known that some test suite are quite buggy themselves (PyTorch comes to mind, but this also applies to LAPACK, see [@bartoldeman's PR](https://github.com/Reference-LAPACK/lapack/pull/764) which dealt with a non-numerical failure in the LAPACK test suite, see also [issue #18017](https://github.com/easybuilders/easybuild-easyconfigs/issues/18017)).

This procedure should probably be documented as well, and even become part of the contribution policy, with some rules of thumb on when this approach is acceptable (how many failing tests we see, on how many CPU targets, etc.).

**Comment by casparvl** [comment]
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Comment by casparvl** [comment]
Created issue for the OpenBLAS test failures:

https://github.com/EESSI/software-layer/issues/314

File: eb_hooks.py [hunk 1]
```diff
@@ -167,7 +167,7 @@ def parse_hook_openblas_relax_lapack_tests_num_errors(ec, eprefix):
         cfg_option = 'max_failing_lapack_tests_num_errors'
         if get_cpu_architecture() == AARCH64:
             orig_value = ec[cfg_option]
-            ec[cfg_option] = 300
+            ec[cfg_option] = 400
```
Comment (boegel) [inline eb_hooks.py:170]: @casparvl Let's add a comment (an extra comment line above) that refers to https://github.com/EESSI/software-layer/issues/314 here, to make the lives of our future selves a bit easier? ;)

Comment (casparvl) [reply eb_hooks.py:170]: Done!

**Pull request merged**